### PR TITLE
INT-3836, INT-4347, INT-4114 - Add organization, group, users, service and enhance property fields

### PR DESCRIPTION
**Pull request by eXtremeX**
- Deconstructed description property of `snyk_finding` entity
- New entities:

  | Resources         | Entity `_type`      | Entity `_class` |
  | ----------------- | ------------------- | --------------- |
  | Snyk Group        | `snyk_group`        | `Group`         |
  | Snyk Organization | `snyk_organization` | `Organization`  |
  | Snyk Service      | `snyk_service`      | `Service`       |
  | Snyk Role         | `snyk_role`         | `AccessRole`    |

- New relationships:

  | Source Entity `_type` | Relationship `_class` | Target Entity `_type` |
  | --------------------- | --------------------- | --------------------- |
  | `snyk_account`        | **HAS**               | `snyk_group`          |
  | `snyk_account`        | **HAS**               | `snyk_service`        |
  | `snyk_group`          | **HAS**               | `snyk_organization`   |
  | `snyk_group`          | **HAS**               | `snyk_role`           |
  | `snyk_organization`   | **HAS**               | `snyk_project`        |
  | `snyk_organization`   | **HAS**               | `snyk_user`           |
  | `snyk_service`        | **IDENTIFIED**        | `snyk_finding`        |
  | `snyk_service`        | **SCANS**             | `snyk_project`        |
  | `snyk_user`           | **ASSIGNED**          | `snyk_role`           |

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,6 +38,8 @@
   },
   "dependencies": {
     "@lifeomic/attempt": "^3.0.3",
+    "@types/lodash.startcase": "^4.4.7",
```
Comment (VDubber) [inline package.json:41]: All @types/* packages should be devDependencies

File: src/config.ts [hunk 2]
```diff
@@ -14,12 +14,16 @@ import {
...
   snykOrgId: string;
   snykApiKey: string;
+  snykGroupId: string;
 }
 
 const instanceConfigFields: IntegrationInstanceConfigFieldMap<IntegrationConfig> = {
   snykOrgId: {
     type: 'string',
   },
+  snykGroupId: {
```
Comment (VDubber) [inline src/config.ts:24]: This is a breaking change. Any thoughts on how to support backwards compatibility?

File: src/converters.ts [hunk 3]
```diff
@@ -7,28 +7,87 @@ import {
...
+        name: data.name,
+      },
+    },
+  });
+}
+
+export function createServiceEntity(service: Service): Entity {
+  return createIntegrationEntity({
+    entityData: {
+      source: {},
```
Comment (VDubber) [inline src/converters.ts:43]: please set `service` as the source

File: src/converters.ts [hunk 4]
```diff
@@ -7,28 +7,87 @@ import {
...
+      assign: {
+        _key: `snyk_service`,
```
Comment (VDubber) [inline src/converters.ts:45]: Is there only every one service? Should the key include the name?

File: src/constants.ts [hunk 5]
```diff
@@ -45,6 +57,21 @@ export const Entities = {
     _class: ['User'],
     resourceName: 'Snyk User',
   },
+  ROLE: {
```
Comment (VDubber) [inline src/constants.ts:60]: Please prefix ROLE, GROUP, ORGANIZATION with SNYK_

File: src/converters.ts [hunk 6]
```diff
@@ -45,6 +104,26 @@ export function getNumericSeverityFromIssueSeverity(
   return numericSeverity === undefined ? 0 : numericSeverity;
 }
 
+export interface Desc {
```
Comment (VDubber) [inline src/converters.ts:107]: nit: rename `Desc` to `Description`?

File: src/converters.ts [hunk 7]
```diff
@@ -45,6 +104,26 @@ export function getNumericSeverityFromIssueSeverity(
...
+  description?: string;
+  Impact?: string;
+  Recommendations?: string;
+  References?: string;
+}
+
+export function deconstructDesc({ desc }: { desc?: string }): Desc | undefined {
```
Comment (VDubber) [inline src/converters.ts:114]: Please add unit tests around this function.

File: src/converters.ts [hunk 8]
```diff
@@ -94,6 +172,25 @@ export function createFindingEntity(vuln: any) {
...
+    },
+  });
+}
+
+export function createRoleEntity(role: Role): Entity {
+  return createIntegrationEntity({
+    entityData: {
+      source: role,
+      assign: {
+        _key: `snyk_role:${role.name.split(' ')[1].toLowerCase()}`,
```
Comment (VDubber) [inline src/converters.ts:186]: Please move to logic and provide unit tests considering edge cases.

File: src/snyk/client.ts [hunk 9]
```diff
@@ -305,4 +309,65 @@ export class APIClient {
       },
     );
   }
+
+  async iterateOrganizations(iteratee: ResourceIteratee<Organization>) {
+    return retry(
```
Comment (VDubber) [inline src/snyk/client.ts:314]: This code is very flawed. It doesn't appear to be introduced in this PR but we should fix it.

Right now the retry logic wraps the request and the iteration logic. If 10 `orgs` are returned and the 8th org is flawed and causes an exception to be thrown, the request would be retried and the same 10 `orgs` would be returned causing a duplicate key error to be thrown. 

I believe the retry logic should be located in its own method. Consider these changes:

```ts
  async snykRequestWithRetry({
    uri,
    method,
    body,
  }: {
    uri: string;
    method: string;
    body?: BodyInit;
  }) {
    return retry(
      () =>
        this.snykRequest({
          uri,
          method,
          body,
        }),
      {
        delay: 5000,
        factor: 1.2,
        maxAttempts: this.retries,
        handleError(err, context) {
          const code = err.statusCode;
          if (code < 500 && code !== 429) {
            context.abort();
          }
        },
      },
    );
  }
```

And then iterateRoles and other methods simply become:

```ts
async listAllProjects(orgId) {
    return this.snykRequestWithRetry({
      method: 'GET',
      uri: `org/${orgId}/projects`,
    });
  }
```

File: src/snyk/client.ts [hunk 10]
```diff
@@ -280,12 +281,15 @@ export class APIClient {
...
+  async iterateUsers(
+    organizationId: string,
+    iteratee: (user: any) => Promise<void>,
+  ) {
     return retry(
       async () => {
         const users = await this.snykRequest({
           method: 'GET',
-          uri: `org/${this.config.snykOrgId}/members?includeGroupAdmins=true`,
+          uri: `org/${organizationId}/members?includeGroupAdmins=true`,
```
Comment (VDubber) [inline src/snyk/client.ts:292]: Is pagination missing from this client?

File: src/config.ts (hunk 2 above)
Comment (VDubber) [reply src/config.ts:24]: I believe we need to use getStepStartStates. We can disable any steps that require the groupId. For an example checkout [graph-github](https://github.com/JupiterOne/graph-github/blob/main/src/getStepStartStates.ts)

File: src/config.ts [hunk 11]
```diff
@@ -31,9 +35,9 @@ export default async function validateInvocation(
 ) {
   const { config } = context.instance;
 
-  if (!config?.snykOrgId || !config?.snykApiKey) {
+  if (!config?.snykOrgId || !config?.snykApiKey || !config?.snykGroupId) {
```
Comment (VDubber) [inline src/config.ts:38]: With the above change, this should not be included in this check for now.

File: src/config.ts (hunk 2 above)
Comment (VDubber) [inline src/config.ts:24]: Is groupid a secret? Should it be masked?

File: src/steps/account/index.ts [hunk 12]
```diff
@@ -4,25 +4,26 @@ import {
...
-  const { instance, jobState, logger } = context;
+async function fetchAccount({
+  instance,
+  jobState,
+}: IntegrationStepExecutionContext<IntegrationConfig>) {
+  const { name, description, id } = instance;
 
-  const apiClient = new APIClient(logger, instance.config);
-  const currentOrgName = await apiClient.getCurrentOrgName();
+  const account: Account = {
```
Comment (VDubber) [inline src/steps/account/index.ts:16]: This is a new pattern for me. Does the provider API not provide desired info for the account?

File: src/steps/fetchFindings.ts [hunk 13]
```diff
@@ -1,31 +1,33 @@
 import {
   createDirectRelationship,
+  Entity,
```
Comment (VDubber) [inline src/steps/fetchFindings.ts:3]: It appears there are no tests for this step, please add some.

File: src/steps/projects/index.ts [hunk 14]
```diff
@@ -1,82 +1,117 @@
...
+          { _key: organizationEntity._key },
+          'Could not get raw data for organization entity',
+        );
+        return;
+      }
+
+      await apiClient.iterateProjects(organization.id, async (project) => {
+        // TODO (austinkelleher, INT-4063) - Remove this later after we've collected some
+        // information on real world origins
+        if (project.origin !== 'github') {
```
Comment (VDubber) [inline src/steps/projects/index.ts:51]: Have we looked at CloudWatch to get some feedback on this?

File: src/steps/users/converter.ts [hunk 15]
```diff
@@ -1,7 +1,8 @@
 import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
 import { Entities } from '../../constants';
+import { User } from '../../types';
 
-export function createUserEntity(user: any) {
+export function createUserEntity(user: User) {
```
Comment (VDubber) [inline src/steps/users/converter.ts:5]: Thank you!

**Review by VDubber — COMMENTED** [review]
I've completed my review. Let me know when it is ready for another review. Thanks @eXtremeX and team!

File: src/steps/account/index.ts (hunk 12 above)
Comment (CreativiceTest) [reply src/steps/account/index.ts:16]: Yep, usually this happens when we need an Entity for the graph of an integration which has no relevant resource in the service - i.e. no endpoint that has the information to build the entity from. We then make entities from information that we manually provide. This happens most often for Service class entities but I've seen it also happen for other classes such as Account, Scanner, and Assessment.

You can find this pattern if you look at the related steps for specs that has [Endpoint: N/A](https://github.com/search?q=org%3AJupiterOne+singleton+endpoint+n%2Fa&type=Code)

Comment (RPGPH) [reply src/steps/account/index.ts:16]: Yep, usually this happens when we need an Entity for the graph of an integration which has no relevant resource in the service - i.e. no endpoint that has the information to build the entity from. We then make entities from information that we manually provide. This happens most often for Service class entities but I've seen it also happen for other classes such as Account, Scanner, and Assessment.

You can find this pattern if you look at the related steps for specs that has [Endpoint: N/A](https://github.com/search?q=org%3AJupiterOne+singleton+endpoint+n%2Fa&type=Code)

File: src/steps/fetchFindings.ts (hunk 13 above)
Comment (RPGPH) [reply src/steps/fetchFindings.ts:3]: Added a test but had to change the step metadata for this.

File: src/converters.ts (hunk 4 above)
Comment (RPGPH) [reply src/converters.ts:47]: Yep, only one service. It's a singleton step 😄

**Comment by RPGPH** [comment]
Hi Samuel, requested changes are pushed. Some of it got a little complicated/counter-intuitive especially the getStepStartStates implementation so feel free to ask me about any clarifications 🙂

File: src/getStepStartStates.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    [StepIds.BUILD_USER_GROUP_ROLE]: { disabled: false },
+    [StepIds.FETCH_PROJECTS]: { disabled: false },
+    [StepIds.FETCH_FINDINGS]: { disabled: false },
+    [StepIds.FETCH_USERS]: { disabled: false },
+    [StepIds.FETCH_ORGANIZATIONS]: { disabled: false },
+
+    [StepIds.BUILD_GROUP_ORG]: { disabled: !isEnterprise },
+    [StepIds.FETCH_GROUP_ROLES]: { disabled: !isEnterprise },
+
+    [StepIds.BUILD_ACCOUNT_ORG]: { disabled: isEnterprise },
```
Comment (zemberdotnet) [inline src/getStepStartStates.ts:27]: Why is `BUILD_ACCOUNT_ORG` scoped by `isEnterprise`? Both the account step and organization step look like they always run.

File: src/util/deconstructDesc.test.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+      'org.jenkins-ci.main:jenkins-core is an open-source automation server. Affected versions of this package are vulnerable to Authentication Bypass. Due to a missing permissions check, a malicious remote authenticated user may be able to trigger the update process of the site.',
+    );
+    expect(Recommendations).toBe(
+      'Upgrade org.jenkins-ci.main:jenkins-core to versions 1.651.2, 2.3 or higher.',
+    );
+    expect(References).toBe('- Jenkins Security Advisory - GitHub Commit');
+    expect(Impact).toBe(
+      'An attacker may be able to trigger the update process of the site',
+    );
+  });
```
Comment (zemberdotnet) [inline src/util/deconstructDesc.test.ts:31]: Great addition 🥳

File: src/steps/organization/index.ts [hunk 18]
```diff
@@ -11,37 +11,90 @@ import { createOrganizationEntity } from '../../converters';
...
-    const organizationEntity = await jobState.addEntity(
-      createOrganizationEntity(organization),
-    );
+  if (groupEntity) {
+    await client.iterateOrganizations(async (org) => {
+      await jobState.addEntity(createOrganizationEntity(org));
+    });
+  } else {
+    const userDetails = await client.getCurrentUserDetails();
+    await jobState.addEntity(createOrganizationEntity(userDetails.orgs[0]));
```
Comment (zemberdotnet) [inline src/steps/organization/index.ts:29]: What if there is multiple orgs?

File: src/steps/group/index.ts [hunk 19]
```diff
@@ -20,26 +20,29 @@ async function fetchGroup({
...
   )) as Entity;
 
-  const groupEntity = await jobState.addEntity(
-    createGroupEntity(await apiClient.getGroupDetails()),
-  );
+  const groupEntity = instance.config.snykGroupId
+    ? await jobState.addEntity(
+        createGroupEntity(await apiClient.getGroupDetails()),
+      )
+    : undefined;
```
Comment (zemberdotnet) [inline src/steps/group/index.ts:27]: This doesn't need to check if `instance.config.snykGroupId` is present. The step should only run when this is present as determined by `getStepStartStates`.

The type for `instance.config.snykGroupId` is possibly undefined though so this is fine.

**Comment by RPGPH** [comment]
Hi Matthew,

I understand that this integration appears to be all over the place. I honestly would like to hear some feedback for this because I found it kinda tricky to implement the integration, specifically with allowing it to support both:

- Enterprise instance: Single group, multiple orgs, multiple projects, and
- Non-enterprise instance: No group, [singular org](https://github.com/JupiterOne/graph-snyk/pull/101#discussion_r1037137444), multiple projects

Let's call them case A and B respectively. 

Intuitively, it seems like it can be implemented by simply utilizing `getStepStartStates` and just disabling the group step when a `groupId` is not provided in the `.env`, but doing it this way will have some problems.

Keeping in mind that in both cases, we have to ingest organization(s),
1. In case A, ingesting organizations is dependent on the group step. **A group has many organizations**.
2. In case B, even if there is no group entity, we should still ingest the **singular organization**. 

If we establish a dependency from the organization step to the group step to implement number 1 (using `dependsOn` step metadata), by disabling the group step in `getStepStartStates` (case B), number 2 will not be satisfied because disabling a step will also disable all steps that are dependent on it.

With this problem established, I thought of two ways to approach it. 

**Run group step for both cases A and B**
1. [Ingest group for case A and store to `jobState`. Case B would leave `jobState` undefined.](https://github.com/JupiterOne/graph-snyk/pull/101#discussion_r1037138128)
2. In the dependent organization step, check for group entity presence in the `jobState`
    - If a group is present, use it to query the group organizations endpoint and ingest all organizations. [Build the relationships](https://github.com/JupiterOne/graph-snyk/pull/101#discussion_r1037128771).
    - If a group is not found, build and ingest the (only) organization entity using information found in the `.env`
3. We proceed with the rest of the integration, no need to touch projects and findings etc.

**Use `getStepStartStates` and run the group step conditionally.**
1. Build two separate steps to ingest organization(s), call them `fetchOrganization` and `fetchGroupOrganizations`.
    - `fetchOrganization` would run when the group step is disabled, build and ingest the (only) organization entity using information found in the `.env`.
    - `fetchGroupOrganizations` would run when the group step is enabled, query the group organizations endpoint and ingest all organizations.
2. Build two separate project steps, and two separate finding steps following the pattern in the previous step.

I chose to go with running the group step for both cases because for the second case, every time we add a step that should run for both cases, _we would need to implement it twice even if both steps are identical_ because of the way dependencies work, like in the current version, project and findings.

The best way to handle this is if we can sort if implement a conditional dependency but I don't think that's possible with the current version of the SDK.

Would love to hear your thoughts on this!

**Comment by eXtremeX** [comment]
Hey @zemberdotnet, looks like there was a bit of misunderstanding:

> I agree with your assessment. Let's implement the first option, running the group step for both cases A and B. I think that will be the simplest.

> I chose to go with running the group step for both cases because for the second case, every time we add a step that should run for both cases, we would need to implement it twice even if both steps are identical because of the way dependencies work, like in the current version, project and findings.

That path is already implemented and today we have rebased it with main (there were conflicts). If you think this is ready, approve it when you can if, if not we're willing to do more changes :toolbox:

File: src/steps/organization/index.ts (hunk 18 above)
Comment (zemberdotnet) [reply src/steps/organization/index.ts:29]: Just a bump on this question.

File: src/steps/roles/converters.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+      source: role,
+      assign: {
+        _key: generateRoleKey(role.name),
+        _type: Entities.SNYK_ROLE._type,
+        _class: Entities.SNYK_ROLE._class,
+        name: role.name,
+        description: role.description,
+        publicId: role.publicId,
+        createdOn: parseTimePropertyValue(role.created),
+        modified: parseTimePropertyValue(role.modified),
```
Comment (zemberdotnet) [inline src/steps/roles/converters.ts:22]: modifiedOn

File: src/util/deconstructDesc.ts [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    if (curr.includes('Remediation'))
+      acc.Recommendations = curr.replace(/Remediation/g, '').trim();
+    else if (curr.includes('References'))
+      acc.References = curr.replace(/References/g, '').trim();
+    else {
+      if (curr.includes('\n\nAn attacker ')) {
+        const splitOverview = curr.split('\n\n');
+        acc.Impact = splitOverview[1].trim();
+        acc.description = splitOverview[0].replace(/Overview/g, '').trim();
+      } else acc.description = curr.replace(/Overview/g, '').trim();
```
Comment (zemberdotnet) [inline src/util/deconstructDesc.ts:23]: Let's make these property names lowercase.

**Review by zemberdotnet — COMMENTED** [review]
A few more things and then we should be good to go @eXtremeX

File: src/steps/organization/index.ts (hunk 18 above)
Comment (RPGPH) [reply src/steps/organization/index.ts]: This else block only gets entered when there is no `groupId provided`. The idea is when that happens, the current behavior should be expected so I refactored it to look like the current implementation. Should be a lot more clearer now.

Here's the referenced code block https://github.com/JupiterOne/graph-snyk/blob/7398d6d4300d9deaef3eb779679cc94bbb0da0e7/src/steps/account/index.ts#L16-L22

**Review by zemberdotnet — APPROVED** [review]

**Comment by eXtremeX** [comment]
@zemberdotnet Had to rebase with main to fix CHANGELOG conflicts, please approve again when you can!

**Review by zemberdotnet — APPROVED** [review]
